src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Storage.scala

```
@@ -0,0 +1,79 @@
+package uk.gov.hmrc.nonrep.attachment
+package service
+
+import java.security.MessageDigest
+
+import akka.actor.typed.ActorSystem
+import akka.http.scaladsl.Http
+import akka.http.scaladsl.client.RequestBuilding.Get
+import akka.http.scaladsl.model.StatusCodes.BadRequest
+import akka.http.scaladsl.model.{HttpRequest, HttpResponse, StatusCodes, Uri}
+import akka.stream.alpakka.s3.scaladsl.S3
+import akka.stream.scaladsl.{Flow, Source}
+import akka.util.ByteString
+import org.apache.commons.codec.digest.DigestUtils
+import uk.gov.hmrc.nonrep.attachment.models.AttachmentRequestKey
+import uk.gov.hmrc.nonrep.attachment.server.ServiceConfig
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+import scala.util.Try
+
+trait Storage[A] {
+  def s3Call(implicit system: ActorSystem[_])
+  : Flow[(HttpRequest, EitherErr[A]), (Try[HttpResponse], EitherErr[A]), Any]
+
+  def s3Request(data: EitherErr[A]): HttpRequest
+
+  def s3Response(request: EitherErr[A], response: HttpResponse)(
+    implicit system: ActorSystem[_]): Future[EitherErr[(A, ByteString)]]
+
+  def putS3Object(attachment: A, file: ByteString)
+                 (implicit system: ActorSystem[_], config: ServiceConfig): Future[EitherErr[A]]
+}
+
+class StorageService extends Storage[AttachmentRequestKey] {
+  import Storage._
+
+  def s3Call(implicit system: ActorSystem[_])
+  : Flow[(HttpRequest, EitherErr[AttachmentRequestKey]), (Try[HttpResponse], EitherErr[AttachmentRequestKey]), Any] =
+    Http().superPool[EitherErr[AttachmentRequestKey]]()
+
+  def s3Request(data: EitherErr[AttachmentRequestKey]): HttpRequest =
+    data.map(attachment => Get(Uri(attachment.request.attachmentUrl))).toOption.getOrElse(HttpRequest())
```

I think it would it be better just to fail fast if for some reason `toOption` returns `None`. 

Maybe something like

`Get(Uri(attachment.request.attachmentUrl))).toOption.getOrElse(throw new RuntimeException("failed to create s3 request"))`

---

You could maybe return the `Either[ErrorMessage, HttpRequest]` but that might get noisy. The exception has the advantage of failing fast while simplifying the return type.